### PR TITLE
[WIP] ADR 040: epic 3

### PR DESCRIPTION
**Pull request by roysc**
## Description

closes: #3

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: db/types.go [hunk 1]
```diff
@@ -91,10 +91,8 @@ type DBReader interface {
...
 
-// A batch/transaction that is also capable of writing to the backing DB.
-type DBReadWriter interface {
-	DBReader
-
+// A write-only batch; can be used to wrap write-optimized batches
+type DBWriter interface {
 	// Set sets the value for the given key, replacing it if it already exists.
 	// CONTRACT: key, value readonly []byte
 	Set([]byte, []byte) error
```
Comment (i-norden) [inline db/types.go:98]: No need for a `Delete` method?

File: store/decoupled/store.go [hunk 2]
```diff
@@ -133,24 +182,26 @@ func (s *Store) Set(key []byte, value []byte) {
...
-	kvHash := s.sc.Get(key)
+	kvHash := s.sckv.Get(key)
 
 	_, err := s.contents().Get(key)
 	if err != nil {
 		panic(err)
 	}
-	s.sc.Delete(key)
+	s.sckv.Delete(key)
 	_ = s.index().Delete(kvHash[:])
```
Comment (i-norden) [inline store/decoupled/store.go:200]: don't need these blank identifiers

File: store/decoupled/store.go [hunk 3]
```diff
@@ -224,12 +284,19 @@ func (s *Store) SetPruning(types.PruningOptions)  {}
...
 
-func (s *Store) versionExists(v uint64) bool {
-	r := s.db.NewReaderAt(v)
+func (s *Store) versionExists(version int64) bool {
+	if version < 0 {
+		return false
+	}
+	r := s.ss.conn.NewReaderAt(uint64(version))
 	return r != nil
+	// for _, valid := range s.ss.Versions() {
```
Comment (i-norden) [inline store/decoupled/store.go]: Can we remove this comment out section?

**Review by i-norden — APPROVED** [review]
LGTM! Just a few minor comments. Also not entirely sure what is going on with the "check annotations" beta features on db/badgerdb/db.go, I think it might be bugged because the package name is "badger" but the import path is  "github.com/dgraph-io/badger/v3"?

File: db/types.go (hunk 1 above)
Comment (roysc) [reply db/types.go:98]: Probably not. I can move it back to the ReadWriter.

edit: Though this actually may be desired if we want to use this for batch operations on the SMT's mapstore

**Comment by roysc** [comment]
Note: this is an outdated version of the ADR-040 work, see the [project board](https://github.com/orgs/vulcanize/projects/5) for links to the current upstream PRs.

**Pull request merged**